keycloak-service-module/src/main/java/org/aerogear/mobile/keycloak_service_module/KeyCloakService.java

```diff
@@ -28,30 +25,34 @@
     private String subjectTokenType;
     private String requestedTokenType;
     private String realm;
-    private MobileCore core;
     private String resource;
     private HttpServiceModule httpModule;
     private String accessToken;
 
-
-    public KeyCloakService() {
+    private KeyCloakService(KeyCloakServiceConfiguration keyCloakServiceConfiguration) {
+        this.serverUrl = keyCloakServiceConfiguration.serverUrl;
+        this.clientId = keyCloakServiceConfiguration.clientId;
+        this.audience = keyCloakServiceConfiguration.audience;
+        this.grantType = keyCloakServiceConfiguration.grantType;
+        this.subjectTokenType = keyCloakServiceConfiguration.subjectTokenType;
+        this.requestedTokenType = keyCloakServiceConfiguration.requestedTokenType;
+        this.realm = keyCloakServiceConfiguration.realm;
+        this.resource = keyCloakServiceConfiguration.resource;
+        this.httpModule = keyCloakServiceConfiguration.httpModule;
     }
 
-
-    @Override
-    public void bootstrap(MobileCore core, ServiceConfiguration config) {
-        this.serverUrl = config.getProperty("auth-server-url");
-        this.clientId = config.getProperty("clientId");
-        this.audience = config.getProperty("audience");
-        this.grantType = config.getProperty("grant_type");
-        this.subjectTokenType = config.getProperty("subject_token_type");
-        this.requestedTokenType = config.getProperty("requested_token_type");
-        this.resource = config.getProperty("resource");
-        this.realm = config.getProperty("realm");
-        this.core = core;
-        this.httpModule = (HttpServiceModule) core.getService("http");
+    public static KeyCloakService getInstance() {
+        ServiceConfiguration config = MobileCore.getServiceConfiguration("keycloak");
+        KeyCloakServiceConfiguration keyCloakServiceConfiguration = new KeyCloakServiceConfiguration(config);
+        return getInstance(keyCloakServiceConfiguration);
     }
 
+    public static KeyCloakService getInstance(KeyCloakServiceConfiguration keyCloakServiceConfiguration) {
```

Comment:
Do we need to make this public? Why would the user want to override the `keyCloakServiceConfiguration` here? This would decouple it from mobile-config.json.

Reply:
@pb82 that is requirement derived from current work. Developers may provide path to certificate files etc and extend configuration.

Reply:
Like @wtrocki said, you should need to add more config not coming from the JSON file. But we can create some sets for it, if people don't agree to have it. @aidenkeating wdyt?

Reply:
@pb82 If we have multiple "keycloak typed" services this won't resolve and there isn't a fallback way to instantiate them by name.  This really doesn't solve the problem, but we will have to have something like the method here in order to do it in this paradigm